Zotero/Controllers/Sync/SyncScheduler.swift

```diff
@@ -8,180 +8,219 @@
-        self.startNextAction()
-    }
-
-    private func enqueueAndStartTimer(action: SchedulerAction) {
-        self.queue.async(flags: .barrier) { [weak self] in
-            self?._enqueueAndStartTimer(action: action)
+    /// Adds a sync to queue. Full sync overrides all queued syncs, since it synchronizes everything, so there's no need to run other syncs before it. Retry sync is scheduled before all syncs
+    /// requested externally, so that the currently running sync can try to finish successfully. Duplicate syncs are not added to queue, no need to run the same sync multiple times in succession.
+    /// - parameter sync: Sync to add to queue
+    private func enqueue(sync: Sync) {
+        if sync.type == .full && sync.libraries == .all {
+            guard self.canPerformFullSync else { return }
+            DDLogInfo("SyncScheduler: clean queue, enqueue full sync")
+            // Full sync overrides all queued syncs, since it will sync up everything
+            self.syncQueue = [sync]
+            // Also reset timer in case we're already waiting for delayed re-sync
+            self.timerDisposeBag = nil
+        } else if self.syncQueue.isEmpty {
+            self.syncQueue.append(sync)
+        } else if sync.retryAttempt > 0 {
+            // Retry sync should be added to the beginning of queue so that retries are processed before new syncs
+            DDLogInfo("SyncScheduler: enqueue retry sync #\(sync.retryAttempt); queue count = \(self.syncQueue.count)")
+            if let index = self.syncQueue.firstIndex(where: { $0.retryAttempt == 0 }) {
+                self.syncQueue.insert(sync, at: index)
+            } else {
+                self.syncQueue.append(sync)
+            }
+        } else if !self.syncQueue.contains(where: { return $0.type == sync.type && $0.libraries == sync.libraries }) {
+            // New sync request should be added to the end of queue if it's not a duplicate
+            self.syncQueue.append(sync)
         }
     }
 
-    private func _enqueueAndStartTimer(action: SchedulerAction) {
-        guard action.syncType != .full || self.canPerformFullSync else { return }
-        self.enqueue(action: action)
-        self.startTimer()
-    }
-
-    private func enqueue(action: SchedulerAction) {
-        guard let (nextSyncType, nextLibrarySyncType) = self.nextAction else {
-            self.nextAction = action
+    /// Start next sync in queue. In case of normal/external syncs wait at least `syncTimeout` between syncs. In case of retry syncs wait for specified delay based on current retry attempt count.
+    private func startNextSync() {
+        // Ignore this if there is a sync in progress. Otherwise pick next sync in queue.
+        guard self.syncInProgress == nil, let nextSync = self.syncQueue.first else {
+            if self.syncInProgress == nil && self.syncQueue.isEmpty {
+                // Report finished queue
+                self.inProgress.accept(false)
+            }
```

Comment:
This seems a bit more clear:
```suggestion
        // Ignore this if there is a sync in progress.
        guard self.syncInProgress == nil else { return }
        // Otherwise pick next sync in queue.
        guard let nextSync = self.syncQueue.first else {
            // Report finished queue
            self.inProgress.accept(false)
```

Reply:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0